### PR TITLE
Story5

### PR DESCRIPTION
**Pull request by fbraese**
this is story 5 (Add a Pet) RFR

### CODE REVIEW

File: client/components/PetAdoptionForm.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
+//aded import to petType page for testing! remove when done!!!
```
Comment (bkincart) [inline client/components/PetAdoptionForm.js]: I don't think this form is supposed to be a part of this PR -- please confirm?

File: client/components/SurrenderForm.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    petName: "",
+    age: "",
+    petTypeId: "",
+    imgUrl: "",
+    vaccinationStatus: "",
+    adoptionStory: "",
+  })
+
+  const handleInputChange = (event) => {
+
```
Comment (bkincart) [inline client/components/SurrenderForm.js:27]: remove empty line

File: client/components/SurrenderForm.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+      [event.currentTarget.id]: event.currentTarget.value,
+    })
+  }
+
+  const submitHandler = (event) => {
+    event.preventDefault()
+    if (isValid()) {
+      addPetForSurrender()
+    } else {
+      console.log("form Not submitted!")
```
Comment (bkincart) [inline client/components/SurrenderForm.js]: Is this the proper error message? the form was submitted, it's that the input wasn't valid, right?

File: client/components/SurrenderForm.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    }
+  }
+
+  const addPetForSurrender = async () => {
+    try {
+      const response = await fetch("/api/v1/surrender", {
```
Comment (bkincart) [inline client/components/SurrenderForm.js]: good! to be a bit more RESTful/conventional, we could make this pluralized `surrenders`

File: client/components/SurrenderForm.js [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    let submitErrors = {}
+    const requiredFields = ['personName', 'phoneNumber', 'email', 'petName', 'age', 'petTypeId', 'imgUrl', 'vaccinationStatus', 'adoptionStory']
+    requiredFields.forEach(field => {
+      if (surrender[field].trim() === "") {
+        submitErrors = {
+          ...submitErrors,
+          [field]: "is blank"
+        }
+      }
+    })
```
Comment (bkincart) [inline client/components/SurrenderForm.js:88]: nicely done :)

File: server/db/schema.sql [hunk 6]
```diff
@@ -17,4 +19,23 @@ CREATE TABLE adoptable_pets(
     adoption_story TEXT NOT NULL,
     available_for_adoption BOOLEAN DEFAULT 't',
     pet_type_id INTEGER REFERENCES pet_types(id)
+);
+
+CREATE TABLE adoption_applications(
```
Comment (bkincart) [inline server/db/schema.sql]: this too looks like it should be in its own PR for Story #4!

File: server/models/Pet.js [hunk 7]
```diff
@@ -5,9 +5,9 @@ const pool = new pg.Pool({
 })
 
 class Pet {
-  constructor({ id, name, age, vaccination_status, vaccinationStatus, adoption_story, adoptionStory, available_for_adoption, availableForAdoption, pet_type_id, petTypeId, type, img_url, imgUrl }) {
+  constructor({ id, name, petName, age, vaccination_status, vaccinationStatus, adoption_story, adoptionStory, available_for_adoption, availableForAdoption, pet_type_id, petTypeId, type, img_url, imgUrl }) {
```
Comment (bkincart) [inline server/models/Pet.js]: We shouldn't be taking in a different name here -- what is the goal? If it's to take a parameter from our Router, the better move would be to hand it in with a key of `name` via the router

File: server/models/Pet.js [hunk 8]
```diff
@@ -43,6 +43,18 @@ class Pet {
...
+    try {
+      const queryString = "INSERT INTO adoptable_pets (name, age, vaccination_status, adoption_story, available_for_adoption, pet_type_id, img_url) VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING id;"
+      const result = await pool.query(queryString, [this.name, this.age, this.vaccinationStatus, this.adoptionStory, this.availableForAdoption, this.petTypeId, this.imgUrl])
+      this.id = result.rows[0].id
+      return this.id
+    } catch (error) {
+      console.error(error)
+      throw error
+    }
+  }
```
Comment (bkincart) [inline server/models/Pet.js:58]: very nice!

File: server/routes/api/v1/surrenderRouter.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import express from "express"
+
+import SurrenderApplication from "../../../models/SurrenderApplication.js"
+
+const surrenderRouter = new express.Router()
+
+surrenderRouter.post("/", async (req, res) => {
+  try {
+    const newSurrender = new SurrenderApplication(req.body)
+    const petId = await newPet.save()
```
Comment (bkincart) [inline server/routes/api/v1/surrenderRouter.js]: where is `newPet` coming from here? This looks like a bug?

File: server/routes/clientRouter.js [hunk 10]
```diff
@@ -2,7 +2,7 @@ import express from "express"
 
 const router = new express.Router()
 
-const clientRoutes = ["/", "/pets", "/pets/:type", "/pets/:type/:id"]
+const clientRoutes = ["/", "/pets", "/pets/:type", "/pets/:type/:id", "/adoptions/new"]
```
Comment (bkincart) [inline server/routes/clientRouter.js:5]: again related to story #4 and will want to remove from this PR

**Review by bkincart — CHANGES_REQUESTED** [review]
Review with changes requested! One big thing: let's get the "New Adoption" stuff into its own branch :)

File: client/components/SurrenderForm.js (hunk 3 above)
Comment (fbraese) [reply client/components/SurrenderForm.js]: Great catch!  We updated to "Inputs not valid!"

File: client/components/SurrenderForm.js (hunk 4 above)
Comment (fbraese) [reply client/components/SurrenderForm.js]: Thank you! Updated

File: server/db/schema.sql (hunk 6 above)
Comment (fbraese) [reply server/db/schema.sql]: Story 4 is part of main

File: server/models/Pet.js (hunk 7 above)
Comment (fbraese) [reply server/models/Pet.js]: It's because our post method takes req.body in for all the information required for a new surrender application and pet at once and pets and surrender applications both have name keys so it is differentiating for the method.  The models for both pet and surrender application are set up to recognized keys for both.  Otherwise the post method may assign human names to pets and vice versa.  Our save surrender application method is set up to take in the pet id from the pet save method so they rely on each other

File: server/routes/clientRouter.js (hunk 10 above)
Comment (fbraese) [reply server/routes/clientRouter.js:5]: Story 4 is in main

File: server/routes/api/v1/surrenderRouter.js (hunk 9 above)
Comment (fbraese) [reply server/routes/api/v1/surrenderRouter.js]: This was on purpose.  The save method for pet returns the petId which is then fed to the save method for the surrender application so the pet primary id and surrender application pet foreign key are the same and reference each other

File: client/components/PetAdoptionForm.js (hunk 1 above)
Comment (fbraese) [reply client/components/PetAdoptionForm.js]: Yep! It's supposed to be here :)

File: server/routes/api/v1/petTypeRouter.js [hunk 11]
```diff
@@ -24,9 +25,13 @@ petTypeRouter.get("/:type", async (req, res) => {
 })
 
 petTypeRouter.get("/:type/:id", async (req, res) => {
+  console.log("inside get with findByTypeAndId method")
   try {
+    console.log(req.params.type)
     const pet = await Pet.findByTypeAndId(req.params.type, req.params.id)
     res.status(200).json({ pet: pet })
+    console.log(pet)
```
Comment (bkincart) [inline server/routes/api/v1/petTypeRouter.js]: Let's remove this log and newline :) as well as the logs on lines 18, 28, and 30!

File: server/routes/api/v1/surrendersRouter.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import express from "express"
+
+import SurrenderApplication from "../../../models/SurrenderApplication.js"
+
+const surrendersRouter = new express.Router()
+
+surrendersRouter.post("/", async (req, res) => {
+  try {
+    const newSurrender = new SurrenderApplication(req.body)
+    const petId = await newPet.save()
```
Comment (bkincart) [inline server/routes/api/v1/surrendersRouter.js:11]: I don't know where my comment from this file went but .. where is `newPet` coming from here? I don't see this variable being created?

File: server/models/Pet.js (hunk 7 above)
Comment (bkincart) [reply server/models/Pet.js]: I would argue that this should then be a concern of the controller, not the model. We don't want to over-extend the signature of the model's constructor by making it _too_ flexible. Instead, in our controller, can we use the spread operator to fill this in correctly?
```
const newPet = new Pet({...req.body, name: req.body.petName})
const newSurrender = new SurrenderApplication({...req.body, name: req.body.personName})
```

**Review by bkincart — CHANGES_REQUESTED** [review]

File: server/models/Pet.js (hunk 7 above)
Comment (fbraese) [reply server/models/Pet.js]: It is also for readability and clarity on the form component since it was confusing having two name keys (one for pet and one for the owner. I hadn't considered using the spread operator in the model like that.  I will try to refactor now

File: server/routes/api/v1/surrendersRouter.js (hunk 12 above)
Comment (fbraese) [reply server/routes/api/v1/surrendersRouter.js:11]: I'm not sure.  It made sense to me when I wrote the code yesterday and then just worked today so I didn't question it.  It's returning a pet and persisting to the database.  I'm adding the new pet declaration now and hoping it doesn't break

File: server/models/Pet.js (hunk 7 above)
Comment (fbraese) [reply server/models/Pet.js]: Okay!  Got the refactor working

File: server/routes/api/v1/surrendersRouter.js (hunk 12 above)
Comment (fbraese) [reply server/routes/api/v1/surrendersRouter.js:11]: Okay!  Refactored to create a pet rather than mysteriously persisting one.

File: client/components/SurrenderForm.js [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+  }
+
+  const submitHandler = (event) => {
+    event.preventDefault()
+    if (isValid()) {
+      addPetForSurrender()
+    } else {
+      console.log("Inputs not valid!")
+    }
```
Comment (safang99) [inline client/components/SurrenderForm.js:54]: since `isValid()` is already adding error messages to `errors` that will be passed to the `ErrorList` component we can remove the `else` part of the condition

File: client/components/SurrenderForm.js [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+          "Content-Type": "application/json",
+        }),
+        body: JSON.stringify(surrender)
+      })
+      if (!response.ok) {
+        const errorMessage = `${response.status} ${response.statusText}`
+        throw new Error(errorMessage)
+      }
+      const responseBody = await response.json()
+      console.log("posted succesfully", responseBody)
```
Comment (safang99) [inline client/components/SurrenderForm.js]: I like seeing this because it let's me know you were checking the data along the way! We can remove it now that development on this portion is over

File: client/components/SurrenderForm.js [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+      setMessage("Your surrender request is in process.")
+      setErrors({})
+      clearForm()
```
Comment (safang99) [inline client/components/SurrenderForm.js:72]: Since `setErrors({})` is already being called from within the `clearForm()` function we can remove 51

File: client/components/SurrenderForm.js [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+            />
+          </label>
+
+          <label htmlFor="petTypeId">Pet Type:
+            <select name="petTypeId" id="petTypeId" onChange={handleInputChange}>
+              <option value=""></option>
+              <option value="2">cat</option>
+              <option value="1">dog </option>
+              <option value="3">mythical creature</option>
+            </select>
```
Comment (safang99) [inline client/components/SurrenderForm.js:173]: Instead of hardcoding this lets do it dynamically! There is already an endpoint that we can set up a useEffect to fetch there which will return all the pet types.  Then we can store that in state and map through it to generate each `<option>` element

**Review by safang99 — CHANGES_REQUESTED** [review]
Looks great! A few small changes with one refactor to make the form dropdown selection dynamic

File: client/components/SurrenderForm.js (hunk 16 above)
Comment (fbraese) [reply client/components/SurrenderForm.js:173]: Great feedback and tips of how to approach this!  We have refactored accordingly
